File: Integrations/DaVinciResolve/321Doit Bridge.lua
-- 321Doit Bridge menu launcher for DaVinci Resolve on macOS.
--
-- Resolve always ships with Lua, while its in-process Python menu support on
-- macOS only detects python.org framework installs. Launch the Python bridge
-- out of process so Homebrew Python works as well. Resolve Studio must allow
-- local external scripting for the Python process to attach.

local home = os.getenv("HOME")
local support = home .. "/Library/Application Support/321Doit/ResolveBridge"
local launcher = support .. "/321Doit Bridge.py"
local log_path = support .. "/launcher.log"

local function file_exists(path)
    local handle = io.open(path, "r")
    if handle then
        handle:close()
        return true
    end
    return false
end

local function shell_quote(value)
    return "'" .. string.gsub(value, "'", "'\\''") .. "'"
end

local python_candidates = {
    "/opt/homebrew/bin/python3",
    "/usr/local/bin/python3",
    "/usr/bin/python3"
}

local python = nil
for _, candidate in ipairs(python_candidates) do
    if file_exists(candidate) then
        python = candidate
        break
    end
end

if not python or not file_exists(launcher) then
    os.execute("/usr/bin/osascript -e " .. shell_quote(
        "display alert \"321Doit Bridge 无法启动\" message \"安装文件不完整，请重新运行安装器。\" as warning"))
    return
end

local api = "/Library/Application Support/Blackmagic Design/DaVinci Resolve/Developer/Scripting"
local lib = "/Applications/DaVinci Resolve/DaVinci Resolve.app/Contents/Libraries/Fusion/fusionscript.so"
local modules = api .. "/Modules"

local command = table.concat({
    "RESOLVE_SCRIPT_API=" .. shell_quote(api),
    "RESOLVE_SCRIPT_LIB=" .. shell_quote(lib),
    "PYTHONPATH=" .. shell_quote(modules),
    shell_quote(python),
    shell_quote(launcher),
    ">>" .. shell_quote(log_path),
    "2>&1 &"
}, " ")

os.execute(command)

File: Integrations/DaVinciResolve/321Doit Bridge.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""321Doit Bridge for DaVinci Resolve — Utility Script entry point.

Run from: Workspace -> Scripts -> 321Doit Bridge
This script only launches the UI window; it never modifies the Resolve
project on startup. All import actions require explicit preflight + execute.
"""

import os
import subprocess
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
# The bridge package is installed outside Resolve's Scripts tree so Resolve
# does not enumerate the internal modules in the Scripts menu.
SUPPORT_DIR = os.path.expanduser(
    "~/Library/Application Support/321Doit/ResolveBridge")
if os.path.isdir(os.path.join(SUPPORT_DIR, "bridge")):
    if SUPPORT_DIR not in sys.path:
        sys.path.insert(0, SUPPORT_DIR)
elif SCRIPT_DIR not in sys.path:
    # Development fallback when running directly from the repository.
    sys.path.insert(0, SCRIPT_DIR)


def _bootstrap_resolve():
    # type: () -> object
    api = os.environ.get(
        "RESOLVE_SCRIPT_API",
        "/Library/Application Support/Blackmagic Design/DaVinci Resolve/Developer/Scripting")
    modules = os.path.join(api, "Modules")
    if modules not in sys.path:
        sys.path.append(modules)
    import DaVinciResolveScript as bmd  # type: ignore
    resolve = bmd.scriptapp("Resolve")
    if resolve is None:
        raise RuntimeError(
            "Resolve 拒绝了本机脚本连接。请在 Resolve 的偏好设置 → "
            "系统 → 常规中，把“外部脚本访问”设为 Local/本地，然后重启 Resolve。")
    return resolve, bmd


def _show_startup_error(message):
    # type: (str) -> None
    sys.stderr.write("321Doit Bridge: %s\n" % message)
    if sys.platform != "darwin":
        return
    escaped = str(message).replace("\\", "\\\\").replace('"', '\\"')
    script = (
        'display alert "321Doit Bridge 无法启动" '
        'message "%s" as warning' % escaped)
    try:
        subprocess.run(
            ["/usr/bin/osascript", "-e", script],
            check=False,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL)
    except Exception:
        pass


def main():
    try:
        resolve, bmd = _bootstrap_resolve()
    except Exception as exc:
        _show_startup_error(str(exc))
        return

    fusion = resolve.Fusion()
    from bridge.ui import run_ui
    run_ui(resolve, fusion, bmd)


if __name__ == "__main__":
    main()
